server/broadcast.go

```diff
@@ -147,6 +144,7 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment, node *core.Live
 				if monitor.Enabled {
 					monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, monitor.SegmentUploadErrorOS, err.Error())
 				}
+				cxn.sessManager.completeSession(sess)
```

Comment:
Pushing the O back on the list after an upload failure may be an issue. Consider the following:

```
O1 uses S3
O2 uses direct upload
S3 goes down
Segment 1 comes in
B selects O1
S3 upload fails
B adds O1 back to beginning of list
Segment 2 comes in
B selects O1
S3 upload fails
B adds O1 back to beginning of list
...
```
By removing O1 from the list outright, we can fall back to O2.

Reply:
You're right - `cxn.sessManager.completeSession(sess)` should be removed from here in the event of consistent OS failures

Reply:
Can be replaced with `cxn.sessManager.removeSession(sess)`